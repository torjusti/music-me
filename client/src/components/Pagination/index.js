import React, { Component } from 'react';
import { connect } from 'react-redux';
import { setPage } from '../../features/pagination/actions';

class Pagination extends Component {
  state = {
    page: this.props.page,
  };

  handleChange = event => {
    this.setState({
      page: event.target.value,
    });
  };

  handleClick = () => {
    this.props.setPage(this.state.page);
  };

  render() {
    return (
      <div>
        <input
          value={this.state.page}
          onChange={this.handleChange}
          type="number"
        />
        <button onClick={this.handleClick}>Go</button>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  page: state.page,
});

const mapDispatchToProps = {
  setPage,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Pagination);
