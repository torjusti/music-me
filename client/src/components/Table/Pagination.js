import React, { Component, Fragment } from 'react';
import { connect } from 'react-redux';
import { setPage } from '../../features/pagination/actions';
import { Icon, Button, Form } from 'semantic-ui-react';
import clamp from 'lodash/clamp';
import styles from './Pagination.module.css';

/**
 * Input box for setting the pagination page manually.
 */
class PageInput extends Component {
  state = {
    page: 1,
  };

  componentDidMount() {
    this.setState({
      page: this.props.pagination.page + 1,
    });
  }

  componentDidUpdate(prevProps) {
    if (this.props !== prevProps) {
      this.setState({
        page: this.props.pagination.page + 1,
      });
    }
  }

  handleChange = event => {
    this.setState({
      page: event.target.value,
    });
  };

  updatePage = () => {
    const current = this.props.pagination.page;

    const page = parseInt(this.state.page, 10) - 1 || current;

    this.props.setPage(clamp(page, 0, this.props.pagination.totalPages - 1));
  };

  handlePress = event => {
    if (event.key === 'Enter') {
      this.updatePage();
    }
  };

  render() {
    return (
      <input
        value={this.state.page}
        onChange={this.handleChange}
        onKeyPress={this.handlePress}
        className={styles.pageInput}
        type="number"
      />
    );
  }
}

/**
 * The two buttons and the display above the table, which
 * allows the user to change the page to show in the table.
 */
export const Pagination = props => {
  const { pagination, setPage } = props;

  return (
    <div className={styles.header}>
      <Button
        className={styles.button}
        onClick={() => setPage(pagination.page - 1)}
        disabled={pagination.page === 0}
      >
        <Icon name="angle left" />
        Prev
      </Button>

      <span className={styles.label}>
        {pagination.totalPages === 0 ? (
          <Fragment>No results</Fragment>
        ) : (
          <Fragment>
            Page <PageInput {...props} /> of {pagination.totalPages}
          </Fragment>
        )}
      </span>

      <Button
        className={styles.button}
        onClick={() => setPage(pagination.page + 1)}
        disabled={
          pagination.page === pagination.totalPages - 1 ||
          pagination.totalPages === 0
        }
      >
        Next
        <Icon name="angle right" />
      </Button>
    </div>
  );
};

const mapStateToProps = state => ({
  pagination: state.pagination,
});

const mapDispatchToProps = {
  setPage,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Pagination);
