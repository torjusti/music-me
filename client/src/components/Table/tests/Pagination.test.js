import React from 'react';
import { mount, shallow } from "enzyme";
import { Provider } from 'react-redux';
import Pagination from '../Pagination';
import configureMockStore from 'redux-mock-store';

describe('Pagination', () => {

  // Create the mock store
  const mockStore = configureMockStore();

  let wrapper, store;

  beforeEach(() => {

    const initialState = {
      pagination: {
        page: 1,
        totalPages: 5
      }
    };

    store = mockStore(initialState);
    wrapper = mount(
      <Provider store={store}>
        <Pagination />
      </Provider>,
    );
  });

  it('renders pagination', () => {
    wrapper = shallow(
      <Pagination store={store}/>,
    );
    expect(wrapper.dive()).toMatchSnapshot();
  });

  it('renders two buttons', () => {
    wrapper = shallow(
      <Pagination store={store} />,
    );
    expect(wrapper.dive().find('Button').length).toEqual(2);
  });
});
