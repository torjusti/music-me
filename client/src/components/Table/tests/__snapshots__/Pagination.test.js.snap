// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination renders pagination 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Pagination
    pagination={
      Object {
        "page": 0,
        "totalPages": 5,
      }
    }
    setPage={[Function]}
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    storeSubscription={
      Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Button
          as="button"
          className="button"
          disabled={true}
          onClick={[Function]}
          role="button"
        >
          <Icon
            as="i"
            name="angle left"
          />
          Prev
        </Button>,
        <span
          className="label"
        >
          <React.Fragment>
            Page 
            <PageInput
              pagination={
                Object {
                  "page": 0,
                  "totalPages": 5,
                }
              }
              setPage={[Function]}
              store={
                Object {
                  "clearActions": [Function],
                  "dispatch": [Function],
                  "getActions": [Function],
                  "getState": [Function],
                  "replaceReducer": [Function],
                  "subscribe": [Function],
                }
              }
              storeSubscription={
                Subscription {
                  "listeners": Object {
                    "clear": [Function],
                    "get": [Function],
                    "notify": [Function],
                    "subscribe": [Function],
                  },
                  "onStateChange": [Function],
                  "parentSub": undefined,
                  "store": Object {
                    "clearActions": [Function],
                    "dispatch": [Function],
                    "getActions": [Function],
                    "getState": [Function],
                    "replaceReducer": [Function],
                    "subscribe": [Function],
                  },
                  "unsubscribe": [Function],
                }
              }
            />
             of 
            5
          </React.Fragment>
        </span>,
        <Button
          as="button"
          className="button"
          disabled={false}
          onClick={[Function]}
          role="button"
        >
          Next
          <Icon
            as="i"
            name="angle right"
          />
        </Button>,
      ],
      "className": "header",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "as": "button",
          "children": Array [
            <Icon
              as="i"
              name="angle left"
            />,
            "Prev",
          ],
          "className": "button",
          "disabled": true,
          "onClick": [Function],
          "role": "button",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "as": "i",
              "name": "angle left",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "Prev",
        ],
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <React.Fragment>
            Page 
            <PageInput
              pagination={
                Object {
                  "page": 0,
                  "totalPages": 5,
                }
              }
              setPage={[Function]}
              store={
                Object {
                  "clearActions": [Function],
                  "dispatch": [Function],
                  "getActions": [Function],
                  "getState": [Function],
                  "replaceReducer": [Function],
                  "subscribe": [Function],
                }
              }
              storeSubscription={
                Subscription {
                  "listeners": Object {
                    "clear": [Function],
                    "get": [Function],
                    "notify": [Function],
                    "subscribe": [Function],
                  },
                  "onStateChange": [Function],
                  "parentSub": undefined,
                  "store": Object {
                    "clearActions": [Function],
                    "dispatch": [Function],
                    "getActions": [Function],
                    "getState": [Function],
                    "replaceReducer": [Function],
                    "subscribe": [Function],
                  },
                  "unsubscribe": [Function],
                }
              }
            />
             of 
            5
          </React.Fragment>,
          "className": "label",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              "Page ",
              <PageInput
                pagination={
                  Object {
                    "page": 0,
                    "totalPages": 5,
                  }
                }
                setPage={[Function]}
                store={
                  Object {
                    "clearActions": [Function],
                    "dispatch": [Function],
                    "getActions": [Function],
                    "getState": [Function],
                    "replaceReducer": [Function],
                    "subscribe": [Function],
                  }
                }
                storeSubscription={
                  Subscription {
                    "listeners": Object {
                      "clear": [Function],
                      "get": [Function],
                      "notify": [Function],
                      "subscribe": [Function],
                    },
                    "onStateChange": [Function],
                    "parentSub": undefined,
                    "store": Object {
                      "clearActions": [Function],
                      "dispatch": [Function],
                      "getActions": [Function],
                      "getState": [Function],
                      "replaceReducer": [Function],
                      "subscribe": [Function],
                    },
                    "unsubscribe": [Function],
                  }
                }
              />,
              " of ",
              5,
            ],
          },
          "ref": null,
          "rendered": Array [
            "Page ",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "pagination": Object {
                  "page": 0,
                  "totalPages": 5,
                },
                "setPage": [Function],
                "store": Object {
                  "clearActions": [Function],
                  "dispatch": [Function],
                  "getActions": [Function],
                  "getState": [Function],
                  "replaceReducer": [Function],
                  "subscribe": [Function],
                },
                "storeSubscription": Subscription {
                  "listeners": Object {
                    "clear": [Function],
                    "get": [Function],
                    "notify": [Function],
                    "subscribe": [Function],
                  },
                  "onStateChange": [Function],
                  "parentSub": undefined,
                  "store": Object {
                    "clearActions": [Function],
                    "dispatch": [Function],
                    "getActions": [Function],
                    "getState": [Function],
                    "replaceReducer": [Function],
                    "subscribe": [Function],
                  },
                  "unsubscribe": [Function],
                },
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            " of ",
            5,
          ],
          "type": Symbol(react.fragment),
        },
        "type": "span",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "as": "button",
          "children": Array [
            "Next",
            <Icon
              as="i"
              name="angle right"
            />,
          ],
          "className": "button",
          "disabled": false,
          "onClick": [Function],
          "role": "button",
        },
        "ref": null,
        "rendered": Array [
          "Next",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "as": "i",
              "name": "angle right",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Button
            as="button"
            className="button"
            disabled={true}
            onClick={[Function]}
            role="button"
          >
            <Icon
              as="i"
              name="angle left"
            />
            Prev
          </Button>,
          <span
            className="label"
          >
            <React.Fragment>
              Page 
              <PageInput
                pagination={
                  Object {
                    "page": 0,
                    "totalPages": 5,
                  }
                }
                setPage={[Function]}
                store={
                  Object {
                    "clearActions": [Function],
                    "dispatch": [Function],
                    "getActions": [Function],
                    "getState": [Function],
                    "replaceReducer": [Function],
                    "subscribe": [Function],
                  }
                }
                storeSubscription={
                  Subscription {
                    "listeners": Object {
                      "clear": [Function],
                      "get": [Function],
                      "notify": [Function],
                      "subscribe": [Function],
                    },
                    "onStateChange": [Function],
                    "parentSub": undefined,
                    "store": Object {
                      "clearActions": [Function],
                      "dispatch": [Function],
                      "getActions": [Function],
                      "getState": [Function],
                      "replaceReducer": [Function],
                      "subscribe": [Function],
                    },
                    "unsubscribe": [Function],
                  }
                }
              />
               of 
              5
            </React.Fragment>
          </span>,
          <Button
            as="button"
            className="button"
            disabled={false}
            onClick={[Function]}
            role="button"
          >
            Next
            <Icon
              as="i"
              name="angle right"
            />
          </Button>,
        ],
        "className": "header",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "as": "button",
            "children": Array [
              <Icon
                as="i"
                name="angle left"
              />,
              "Prev",
            ],
            "className": "button",
            "disabled": true,
            "onClick": [Function],
            "role": "button",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "as": "i",
                "name": "angle left",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "Prev",
          ],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <React.Fragment>
              Page 
              <PageInput
                pagination={
                  Object {
                    "page": 0,
                    "totalPages": 5,
                  }
                }
                setPage={[Function]}
                store={
                  Object {
                    "clearActions": [Function],
                    "dispatch": [Function],
                    "getActions": [Function],
                    "getState": [Function],
                    "replaceReducer": [Function],
                    "subscribe": [Function],
                  }
                }
                storeSubscription={
                  Subscription {
                    "listeners": Object {
                      "clear": [Function],
                      "get": [Function],
                      "notify": [Function],
                      "subscribe": [Function],
                    },
                    "onStateChange": [Function],
                    "parentSub": undefined,
                    "store": Object {
                      "clearActions": [Function],
                      "dispatch": [Function],
                      "getActions": [Function],
                      "getState": [Function],
                      "replaceReducer": [Function],
                      "subscribe": [Function],
                    },
                    "unsubscribe": [Function],
                  }
                }
              />
               of 
              5
            </React.Fragment>,
            "className": "label",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": Array [
                "Page ",
                <PageInput
                  pagination={
                    Object {
                      "page": 0,
                      "totalPages": 5,
                    }
                  }
                  setPage={[Function]}
                  store={
                    Object {
                      "clearActions": [Function],
                      "dispatch": [Function],
                      "getActions": [Function],
                      "getState": [Function],
                      "replaceReducer": [Function],
                      "subscribe": [Function],
                    }
                  }
                  storeSubscription={
                    Subscription {
                      "listeners": Object {
                        "clear": [Function],
                        "get": [Function],
                        "notify": [Function],
                        "subscribe": [Function],
                      },
                      "onStateChange": [Function],
                      "parentSub": undefined,
                      "store": Object {
                        "clearActions": [Function],
                        "dispatch": [Function],
                        "getActions": [Function],
                        "getState": [Function],
                        "replaceReducer": [Function],
                        "subscribe": [Function],
                      },
                      "unsubscribe": [Function],
                    }
                  }
                />,
                " of ",
                5,
              ],
            },
            "ref": null,
            "rendered": Array [
              "Page ",
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "pagination": Object {
                    "page": 0,
                    "totalPages": 5,
                  },
                  "setPage": [Function],
                  "store": Object {
                    "clearActions": [Function],
                    "dispatch": [Function],
                    "getActions": [Function],
                    "getState": [Function],
                    "replaceReducer": [Function],
                    "subscribe": [Function],
                  },
                  "storeSubscription": Subscription {
                    "listeners": Object {
                      "clear": [Function],
                      "get": [Function],
                      "notify": [Function],
                      "subscribe": [Function],
                    },
                    "onStateChange": [Function],
                    "parentSub": undefined,
                    "store": Object {
                      "clearActions": [Function],
                      "dispatch": [Function],
                      "getActions": [Function],
                      "getState": [Function],
                      "replaceReducer": [Function],
                      "subscribe": [Function],
                    },
                    "unsubscribe": [Function],
                  },
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              " of ",
              5,
            ],
            "type": Symbol(react.fragment),
          },
          "type": "span",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "as": "button",
            "children": Array [
              "Next",
              <Icon
                as="i"
                name="angle right"
              />,
            ],
            "className": "button",
            "disabled": false,
            "onClick": [Function],
            "role": "button",
          },
          "ref": null,
          "rendered": Array [
            "Next",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "as": "i",
                "name": "angle right",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`Pagination should send an action when pressing the next button 1`] = `
Array [
  Object {
    "payload": Object {
      "page": 1,
    },
    "type": "SET_PAGE",
  },
]
`;

exports[`Pagination should send an action when pressing the previous button 1`] = `Array []`;
